### ISSUE TITLE
Problem: when no_pin is given endstop is triggered but no error is given 

### ISSUE THREAD

**Issue by erikcramerjr** [bdring/FluidNC#989]
### Wiki Search Terms

Six pack, no pin  

### Controller Board

6-pack

### Machine Description

XYYZ homing on all axis  TMC5160 Drivers

### Configuration file

```yaml
motor0:
      limit_neg_pin: gpio.4:high
      limit_pos_pin: NO_PIN
      limit_all_pin: NO_PIN

    motor0:
      limit_neg_pin: gpio.4:high
      limit_pos_pin: NO_PIN
      limit_all_pin: 


board: 6 Pack
name: 6 Pack TMC5160  XYYZ homing
stepping:
  engine: I2S_STREAM
  idle_ms: 250
  pulse_us: 4
  dir_delay_us: 1
  disable_delay_us: 0

axes:
  shared_stepper_disable_pin: NO_PIN
  x:
    steps_per_mm: 800.000
    max_rate_mm_per_min: 5000.000
    acceleration_mm_per_sec2: 100.000
    max_travel_mm: 300.000
    soft_limits: false
    homing:
      cycle: 2
      positive_direction: false
      mpos_mm: 150.000
      feed_mm_per_min: 100.000
      seek_mm_per_min: 200.000
      settle_ms: 500
      seek_scaler: 1.100
      feed_scaler: 1.100

    motor0:
      limit_neg_pin: gpio.27:high
      limit_pos_pin: NO_PIN
      limit_all_pin: NO_PIN
      hard_limits: false
      pulloff_mm: 1.000
      tmc_5160:
                      
        r_sense_ohms: 0.110
        run_amps: 0.75
        hold_amps: 0.50
        microsteps: 32
        stallguard: 0
        stallguard_debug: false
        toff_disable: 0
        toff_stealthchop: 5
        toff_coolstep: 3
        run_mode: StealthChop
        homing_mode: StealthChop
        use_enable: false
        cs_pin: i2so.3
        step_pin: I2SO.2
        direction_pin: I2SO.1
        disable_pin: I2SO.0

  y:
    steps_per_mm: 800.000
    max_rate_mm_per_min: 5000.000
    acceleration_mm_per_sec2: 100.000
    max_travel_mm: 300.000
    soft_limits: false
    homing:
      cycle: 2
      positive_direction: true
      mpos_mm: 150.000
      feed_mm_per_min: 100.000
      seek_mm_per_min: 200.000
      settle_ms: 500
      seek_scaler: 1.100
      feed_scaler: 1.100

    motor0:
      limit_neg_pin: gpio.4:high
      limit_pos_pin: NO_PIN
      limit_all_pin: NO_PIN
      hard_limits: false
      pulloff_mm: 1.000
      tmc_5160:
        cs_pin: i2so.6
        r_sense_ohms: 0.110
        run_amps: 0.75
        hold_amps: 0.50
        microsteps: 32
        stallguard: 0
        stallguard_debug: false
        toff_disable: 0
        toff_stealthchop: 5
        toff_coolstep: 3
        run_mode: StealthChop
        homing_mode: StealthChop
        use_enable: false
        step_pin: I2SO.5
        direction_pin: I2SO.4
        disable_pin: I2SO.7

    motor1:
      limit_neg_pin: NO_PIN
      limit_pos_pin: NO_PIN
      limit_all_pin: NO_PIN
      hard_limits: false
      pulloff_mm: 1.000
      tmc_5160:
        cs_pin: i2so.11
        r_sense_ohms: 0.110
        run_amps: 0.750
        hold_amps: 0.50
        microsteps: 32
        stallguard: 0
        stallguard_debug: false
        toff_disable: 0
        toff_stealthchop: 5
        toff_coolstep: 3
        run_mode: StealthChop
        homing_mode: StealthChop
        use_enable: false
        step_pin: I2SO.10
        direction_pin: I2SO.9
        disable_pin: I2SO.8

  z:
    steps_per_mm: 800.000
    max_rate_mm_per_min: 5000.000
    acceleration_mm_per_sec2: 100.000
    max_travel_mm: 300.000
    soft_limits: false
    homing:
      cycle: 1
      positive_direction: true
      mpos_mm: 150.000
      feed_mm_per_min: 100.000
      seek_mm_per_min: 800.000
      settle_ms: 500
      seek_scaler: 1.100
      feed_scaler: 1.100

    motor0:
      limit_neg_pin: gpio.16:high
      limit_pos_pin: NO_PIN
      limit_all_pin: NO_PIN
      hard_limits: false
      pulloff_mm: 1.000
      tmc_5160:
        cs_pin: i2so.14
        r_sense_ohms: 0.110
        run_amps: 0.750
        hold_amps: 0.50
        microsteps: 32
        stallguard: 0
        stallguard_debug: false
        toff_disable: 0
        toff_stealthchop: 5
        toff_coolstep: 3
        run_mode: StealthChop
        homing_mode: StealthChop
        use_enable: false
        step_pin: I2SO.13
        direction_pin: I2SO.12
        disable_pin: I2SO.15
i2c0:
   sda_pin: gpio.14
   scl_pin: gpio.13
        
oled:
   i2c_num: 0
   i2c_address: 60
   width: 128
   height: 64
   radio_delay_ms: 1000
i2so:
  bck_pin: gpio.22
  data_pin: gpio.21
  ws_pin: gpio.17

spi:
  miso_pin: gpio.19
  mosi_pin: gpio.23
  sck_pin: gpio.18

sdcard:
  card_detect_pin: NO_PIN
  cs_pin: gpio.5
  frequency_hz: 4000000 
control:
  safety_door_pin: NO_PIN
  reset_pin: NO_PIN
  feed_hold_pin: NO_PIN
  cycle_start_pin: NO_PIN
  macro0_pin: NO_PIN
  macro1_pin: NO_PIN
  macro2_pin: NO_PIN
  macro3_pin: NO_PIN



user_outputs:
  analog0_pin: NO_PIN
  analog1_pin: NO_PIN
  analog2_pin: NO_PIN
  analog3_pin: NO_PIN
  analog0_hz: 5000
  analog1_hz: 5000
  analog2_hz: 5000
  analog3_hz: 5000
  digital0_pin: NO_PIN
  digital1_pin: NO_PIN
  digital2_pin: NO_PIN
  digital3_pin: NO_PIN

start:
  must_home: false

 
and : 
board: 6 Pack
name: 6 Pack TMC5160  XYYZ homing
stepping:
  engine: I2S_STREAM
  idle_ms: 250
  pulse_us: 4
  dir_delay_us: 1
  disable_delay_us: 0

axes:
  shared_stepper_disable_pin: NO_PIN
  x:
    steps_per_mm: 800.000
    max_rate_mm_per_min: 5000.000
    acceleration_mm_per_sec2: 100.000
    max_travel_mm: 300.000
    soft_limits: false
    homing:
      cycle: 2
      positive_direction: false
      mpos_mm: 150.000
      feed_mm_per_min: 100.000
      seek_mm_per_min: 200.000
      settle_ms: 500
      seek_scaler: 1.100
      feed_scaler: 1.100

    motor0:
      limit_neg_pin: gpio.27:high
      limit_pos_pin: NO_PIN
      limit_all_pin: 
      hard_limits: false
      pulloff_mm: 1.000
      tmc_5160:
                      
        r_sense_ohms: 0.110
        run_amps: 0.75
        hold_amps: 0.50
        microsteps: 32
        stallguard: 0
        stallguard_debug: false
        toff_disable: 0
        toff_stealthchop: 5
        toff_coolstep: 3
        run_mode: StealthChop
        homing_mode: StealthChop
        use_enable: false
        cs_pin: i2so.3
        step_pin: I2SO.2
        direction_pin: I2SO.1
        disable_pin: I2SO.0

  y:
    steps_per_mm: 800.000
    max_rate_mm_per_min: 5000.000
    acceleration_mm_per_sec2: 100.000
    max_travel_mm: 300.000
    soft_limits: false
    homing:
      cycle: 2
      positive_direction: true
      mpos_mm: 150.000
      feed_mm_per_min: 100.000
      seek_mm_per_min: 200.000
      settle_ms: 500
      seek_scaler: 1.100
      feed_scaler: 1.100

    motor0:
      limit_neg_pin: gpio.4:high
      limit_pos_pin: NO_PIN
      limit_all_pin: NO_PIN
      hard_limits: false
      pulloff_mm: 1.000
      tmc_5160:
        cs_pin: i2so.6
        r_sense_ohms: 0.110
        run_amps: 0.75
        hold_amps: 0.50
        microsteps: 32
        stallguard: 0
        stallguard_debug: false
        toff_disable: 0
        toff_stealthchop: 5
        toff_coolstep: 3
        run_mode: StealthChop
        homing_mode: StealthChop
        use_enable: false
        step_pin: I2SO.5
        direction_pin: I2SO.4
        disable_pin: I2SO.7

    motor1:
      limit_neg_pin: NO_PIN
      limit_pos_pin: NO_PIN
      limit_all_pin: NO_PIN
      hard_limits: false
      pulloff_mm: 1.000
      tmc_5160:
        cs_pin: i2so.11
        r_sense_ohms: 0.110
        run_amps: 0.750
        hold_amps: 0.50
        microsteps: 32
        stallguard: 0
        stallguard_debug: false
        toff_disable: 0
        toff_stealthchop: 5
        toff_coolstep: 3
        run_mode: StealthChop
        homing_mode: StealthChop
        use_enable: false
        step_pin: I2SO.10
        direction_pin: I2SO.9
        disable_pin: I2SO.8

  z:
    steps_per_mm: 800.000
    max_rate_mm_per_min: 5000.000
    acceleration_mm_per_sec2: 100.000
    max_travel_mm: 300.000
    soft_limits: false
    homing:
      cycle: 1
      positive_direction: true
      mpos_mm: 150.000
      feed_mm_per_min: 100.000
      seek_mm_per_min: 800.000
      settle_ms: 500
      seek_scaler: 1.100
      feed_scaler: 1.100

    motor0:
      limit_neg_pin: gpio.16:high
      limit_pos_pin: NO_PIN
      limit_all_pin: NO_PIN
      hard_limits: false
      pulloff_mm: 1.000
      tmc_5160:
        cs_pin: i2so.14
        r_sense_ohms: 0.110
        run_amps: 0.750
        hold_amps: 0.50
        microsteps: 32
        stallguard: 0
        stallguard_debug: false
        toff_disable: 0
        toff_stealthchop: 5
        toff_coolstep: 3
        run_mode: StealthChop
        homing_mode: StealthChop
        use_enable: false
        step_pin: I2SO.13
        direction_pin: I2SO.12
        disable_pin: I2SO.15
i2c0:
   sda_pin: gpio.14
   scl_pin: gpio.13
        
oled:
   i2c_num: 0
   i2c_address: 60
   width: 128
   height: 64
   radio_delay_ms: 1000
i2so:
  bck_pin: gpio.22
  data_pin: gpio.21
  ws_pin: gpio.17

spi:
  miso_pin: gpio.19
  mosi_pin: gpio.23
  sck_pin: gpio.18

sdcard:
  card_detect_pin: NO_PIN
  cs_pin: gpio.5
  frequency_hz: 4000000 
control:
  safety_door_pin: NO_PIN
  reset_pin: NO_PIN
  feed_hold_pin: NO_PIN
  cycle_start_pin: NO_PIN
  macro0_pin: NO_PIN
  macro1_pin: NO_PIN
  macro2_pin: NO_PIN
  macro3_pin: NO_PIN



user_outputs:
  analog0_pin: NO_PIN
  analog1_pin: NO_PIN
  analog2_pin: NO_PIN
  analog3_pin: NO_PIN
  analog0_hz: 5000
  analog1_hz: 5000
  analog2_hz: 5000
  analog3_hz: 5000
  digital0_pin: NO_PIN
  digital1_pin: NO_PIN
  digital2_pin: NO_PIN
  digital3_pin: NO_PIN

start:
  must_home: false
```

### Startup Messages

```text
[MSG:INFO: FluidNC v3.7.5]
[MSG:INFO: Compiled with ESP32 SDK:v4.4.4]
[MSG:DBG: Spiffs mount failed: ESP_FAIL]
[MSG:INFO: Local filesystem type is littlefs]
[MSG:INFO: Configuration file:6_pack_TMC5160_XYYZv14.yaml]
[MSG:DBG: Running after-parse tasks]
[MSG:DBG: Checking configuration]
[MSG:INFO: Machine 6 Pack TMC5160  XYYZ homing]
[MSG:INFO: Board 6 Pack]
[MSG:INFO: I2SO BCK:gpio.22 WS:gpio.17 DATA:gpio.21]
[MSG:INFO: SPI SCK:gpio.18 MOSI:gpio.23 MISO:gpio.19]
[MSG:INFO: SD Card cs_pin:gpio.5 detect:NO_PIN freq:4000000]
[MSG:INFO: I2C SDA: gpio.14, SCL: gpio.13, Freq: 100000, Bus #: 0]
[MSG:INFO: OLED I2C address:0x3c width: 128 height: 64]
[MSG:INFO: Stepping:I2S_stream Pulse:4us Dsbl Delay:0us Dir Delay:1us Idle Delay:250ms]
[MSG:INFO: Axis count 3]
[MSG:INFO: Axis X (150.000,450.000)]
[MSG:INFO:   Motor0]
[MSG:DBG: TMCStepper Library Ver. 0x0x703]
[MSG:INFO:     tmc_5160 Step:I2SO.2 Dir:I2SO.1 CS:I2SO.3 Disable:I2SO.0 Index:-1 R:0.110]
[MSG:INFO:  X Neg Limit gpio.27]
[MSG:DBG:  X Neg Limit 0]
[MSG:INFO:  X All Limit NO_PIN]
[MSG:DBG:  X All Limit 0]
[MSG:INFO: Axis Y (-150.000,150.000)]
[MSG:INFO:   Motor0]
[MSG:INFO:     tmc_5160 Step:I2SO.5 Dir:I2SO.4 CS:I2SO.6 Disable:I2SO.7 Index:-1 R:0.110]
[MSG:INFO:  Y Neg Limit gpio.4]
[MSG:DBG:  Y Neg Limit 0]
[MSG:INFO:   Motor1]
[MSG:INFO:     tmc_5160 Step:I2SO.10 Dir:I2SO.9 CS:I2SO.11 Disable:I2SO.8 Index:-1 R:0.110]
[MSG:INFO: Axis Z (-150.000,150.000)]
[MSG:INFO:   Motor0]
[MSG:INFO:     tmc_5160 Step:I2SO.13 Dir:I2SO.12 CS:I2SO.14 Disable:I2SO.15 Index:-1 R:0.110]
[MSG:INFO:  Z Neg Limit gpio.16]
[MSG:DBG:  Z Neg Limit 0]
[MSG:INFO: X Axis driver test passed]
[MSG:DBG: X Axis StealthChop]
[MSG:DBG: CHOPCONF: 0x0x13408158]
[MSG:DBG: COOLCONF: 0x0x0]
[MSG:DBG: THIGH: 0x0x0]
[MSG:DBG: TCOOLTHRS: 0x0x0]
[MSG:DBG: GCONF: 0x0x4]
[MSG:DBG: PWMCONF: 0x0xc40c001e]
[MSG:DBG: IHOLD_IRUN: 0x0x150e]
[MSG:INFO: Y Axis driver test passed]
[MSG:DBG: Y Axis StealthChop]
[MSG:DBG: CHOPCONF: 0x0x13408158]
[MSG:DBG: COOLCONF: 0x0x0]
[MSG:DBG: THIGH: 0x0x0]
[MSG:DBG: TCOOLTHRS: 0x0x0]
[MSG:DBG: GCONF: 0x0x4]
[MSG:DBG: PWMCONF: 0x0xc40c001e]
[MSG:DBG: IHOLD_IRUN: 0x0x150e]
[MSG:INFO: Y2 Axis driver test passed]
[MSG:DBG: Y2 Axis StealthChop]
[MSG:DBG: CHOPCONF: 0x0x13408158]
[MSG:DBG: COOLCONF: 0x0x0]
[MSG:DBG: THIGH: 0x0x0]
[MSG:DBG: TCOOLTHRS: 0x0x0]
[MSG:DBG: GCONF: 0x0x4]
[MSG:DBG: PWMCONF: 0x0xc40c001e]
[MSG:DBG: IHOLD_IRUN: 0x0x150e]
[MSG:INFO: Z Axis driver test passed]
[MSG:DBG: Z Axis StealthChop]
[MSG:DBG: CHOPCONF: 0x0x13408158]
[MSG:DBG: COOLCONF: 0x0x0]
[MSG:DBG: THIGH: 0x0x0]
[MSG:DBG: TCOOLTHRS: 0x0x0]
[MSG:DBG: GCONF: 0x0x4]
[MSG:DBG: PWMCONF: 0x0xc40c001e]
[MSG:DBG: IHOLD_IRUN: 0x0x150e]
[MSG:INFO: Kinematic system: Cartesian]
[MSG:INFO: Using spindle NoSpindle]
[MSG:INFO: Connecting to STA SSID:yourwifihere]
[MSG:INFO: Connecting.]
[MSG:INFO: Connecting..]
[MSG:INFO: Connected - IP is 192.XXX]
[MSG:INFO: WiFi on]
[MSG:INFO: Start mDNS with hostname:http://fluidnc.local/]
[MSG:INFO: SSDP Started]
[MSG:INFO: HTTP started on port 80]
[MSG:DBG: 6_pack_TMC5160_XYYZv12.yaml hash "475370808518B360B9AFFA3BC9785AA6722A1B170EB0A10075E490C2103F431E"]
[MSG:DBG: 6_pack_TMC5160_XYYZv13.yaml hash "3CE4297A05DCA03E0ACA62AA3A23CBC2D28A8B3AC3512B4C13344B5218307B92"]
[MSG:DBG: 6_pack_TMC5160_XYYZv14.yaml hash "6DA679A096E54EA9E80B190508059AE87580BF84CEC64DA73E436DA3E76D0299"]
[MSG:DBG: 6_pack_TMC5160_XYYZv6.yaml hash "C717CC9890C898902D1F345F22B5CB25B2F0984BC13C3B96616E4F635BB2613F"]
[MSG:DBG: 6_pack_TMC5160_XYYZv7.yaml hash "5E89FC0E69515FA04E667BBD4E4A3C47545BA78A1A05E4800B093841F33A39DD"]
[MSG:DBG: favicon.ico hash "B8B2871A343CA0F9A7A130213226962AD2BE4DFD28D7D8B756C3557569CB876D"]
[MSG:DBG: index.html.gz hash "94B77CC427B19E67041AC4B3891318595E3B4D0E1337479D1A0DC8F51244A6A6"]
[MSG:DBG: preferences.json hash "E2D32AF149AB3E43CADBAE0F7E353F4C65E150837E023EABD8C474DFF3A11203"]
[MSG:INFO: Telnet started on port 23]

Grbl 3.7 [FluidNC v3.7.5 (wifi) '$' for help]
[MSG:DBG:  X Neg Limit 0]
[MSG:DBG:  Z Neg Limit 0]
[MSG:DBG:  Y Neg Limit 0]
[MSG:DBG:  X All Limit 1]
[MSG:DBG: Limit switch tripped for X motor 0]
[MSG:DBG:  X Neg Limit 1]
[MSG:DBG: Limit switch tripped for X motor 0]
[MSG:DBG:  X Neg Limit 0]
[MSG:DBG:  X Neg Limit 1]
[MSG:DBG: Limit switch tripped for X motor 0]
[MSG:DBG:  X Neg Limit 0]
[MSG:DBG:  X Neg Limit 1]
[MSG:DBG: Limit switch tripped for X motor 0]
[MSG:DBG:  X Neg Limit 0]
[MSG:DBG: index.html is cached]
[MSG:DBG: /favicon.ico is cached]
[MSG:DBG: WebSocket 0 from 192.XXX uri /]
[MSG:DBG: /macrocfg.json not found]
[MSG:DBG: 404.htm not found]
[MSG:DBG: /preferences.json is cached]
[MSG:DBG:  Y Neg Limit 1]
[MSG:DBG: Limit switch tripped for Y motor 0]
[MSG:DBG:  Y Neg Limit 0]
[MSG:DBG:  Y Neg Limit 1]
[MSG:DBG: Limit switch tripped for Y motor 0]
[MSG:DBG:  Y Neg Limit 0]
[MSG:DBG:  Y Neg Limit 1]
[MSG:DBG: Limit switch tripped for Y motor 0]
[MSG:DBG:  Y Neg Limit 0]
[MSG:DBG:  Y Neg Limit 1]
[MSG:DBG: Limit switch tripped for Y motor 0]
[MSG:DBG:  Y Neg Limit 0]
```

### What happened?

When you don't set one of the pins, and don't set NO-PIN, no error is given during start-up but the endstop for the motor is set to high automatically. What i would expect that would happen is a error at start up that the config is not good, same as when you use a pin twice.

  motor0:
      limit_neg_pin: gpio.4:high
      limit_pos_pin: NO_PIN
      limit_all_pin: NO_PIN
VS

    motor0:
      limit_neg_pin: gpio.4:high
      limit_pos_pin: NO_PIN
      limit_all_pin: 

### GCode File

no gcode used

### Other Information

.

**Comment by bdring** [comment]
# 1

I don't understand your "What Happened?"

# 2

If I start a config file like yours ...

```yaml
motor0:
      limit_neg_pin: gpio.4:high
      limit_pos_pin: NO_PIN
      limit_all_pin: NO_PIN

    motor0:
      limit_neg_pin: gpio.4:high
      limit_pos_pin: NO_PIN
      limit_all_pin: 


board: 6 Pack
name: 6 Pack TMC5160  XYYZ homing
...
```

I get a bunch of errors that you don't. You must show start messages from the included config or I am wasting my time.

```
[MSG:WARN: Ignored key motor0]
[MSG:ERR: Skipping key limit_neg_pin indent 6 this indent 0]
[MSG:ERR: Skipping key limit_pos_pin indent 6 this indent 0]
[MSG:ERR: Skipping key limit_all_pin indent 6 this indent 0]
[MSG:ERR: Skipping key motor0 indent 4 this indent 0]
[MSG:ERR: Skipping key limit_neg_pin indent 6 this indent 0]
[MSG:ERR: Skipping key limit_pos_pin indent 6 this indent 0]
[MSG:ERR: Skipping key limit_all_pin indent 6 this indent 0]
[MSG:INFO: Machine 6 Pack StepStick XYZ]
[MSG:INFO: Board 6 Pack]
```
# 3

If you don't set a pin to NO_PIN it will do it for you.

This ...

```
    motor0:
      limit_neg_pin: NO_PIN
      limit_pos_pin: NO_PIN
      limit_all_pin: 
```

is the same as this ...

```
    motor0:
      limit_neg_pin: NO_PIN
      limit_pos_pin: NO_PIN
```
Both result in this

```
  motor0:
    limit_neg_pin: NO_PIN
    limit_pos_pin: NO_PIN
    limit_all_pin: NO_PIN
```

This is true for all pins that are not required.

I use the PWM spindle for example

```
pwm:
  output_pin: NO_PIN
```

or 

```
pwm:
  output_pin: NO_PIN
```
will give a error like this and force a default config.

```
[MSG:ERR: PWM output pin not defined]
```

# 4

Pins can have attributes. If you do not provide them defaults will be used. They is active high and no pull up or pull down.

**Comment by MitchBradley** [comment]
Fixed by #991

**Comment by erikcramerjr** [comment]
Do you still need the start messages, or can i mark it as fixed ?
